### PR TITLE
WIP - thiatt/pulumi/content/challenge/startup-in-a-box/gcp

### PR DESCRIPTION
**Pull request by timbohiatt**
WIP - Pulumi Startup in a Box Challenge

- Add Challenge Code
- Ensure tutorial is formatted with Pulumi Style and Pass lint checks. 
- Separating out AWS and GCP Versions of the challenge.

### CODE REVIEW

File: themes/default/content/challenge/startup-in-a-box.md [hunk 1]
```diff
@@ -14,32 +14,37 @@ meta_image: /images/challenge/challenge_cta.png
...
     <img class="mx-auto my-4 md:hidden" src="/images/pulumi-challenge-swag-tumbler-mobile.png" />
   </div>
 </div>
 
+{{< chooser cloud "aws,gcp" >}}
+
+{{% choosable cloud aws %}}
+<div class="w-full md:w-1/2">
+  <h3>Prerequisites</h3>
+  <p>In order to complete this challenge, you'll need a couple things set up in advance.</p>
```
Comment (gangchen03) [inline themes/default/content/challenge/startup-in-a-box.md]: minor correction: `you'll need a couple of things set up in advance.`

File: themes/default/content/challenge/startup-in-a-box.md [hunk 2]
```diff
@@ -508,9 +513,410 @@ expect(await page.title()).toBe("Pulumi Challenge");
 await browser.close();
 ```
 
+{{% /choosable %}}
+
+{{% choosable cloud gcp %}}
+
+<div class="w-full md:w-1/2">
+  <h3>Prerequisites</h3>
+  <p>In order to complete this challenge, you'll need a couple things set up in advance.</p>
```
Comment (gangchen03) [inline themes/default/content/challenge/startup-in-a-box.md]: `a couple of things`

File: themes/default/content/challenge/startup-in-a-box.md [hunk 3]
```diff
@@ -508,9 +513,410 @@ expect(await page.title()).toBe("Pulumi Challenge");
...
+
+### Challenge
+
+Follow along with the steps outlined on this page to complete the challenge.
+
+#### Step 1. Your First Pulumi Program
+
+You will learn how to create a new Pulumi program using our Pulumi templates, specifically for GCP with TypeScript. Create a new directory called `pulumi-challenge` and run the following inside of it:
+
+```shell
```
Comment (gangchen03) [inline themes/default/content/challenge/startup-in-a-box.md:549]: Better add the folder creation commands as well:
```
mkdir pulumi-challenge
cd pulumi-challenge
```

File: themes/default/content/challenge/startup-in-a-box.md [hunk 4]
```diff
@@ -508,9 +513,410 @@ expect(await page.title()).toBe("Pulumi Challenge");
...
+  # set up environment
+  pulumi new gcp-typescript
```
Comment (gangchen03) [inline themes/default/content/challenge/startup-in-a-box.md:551]: The tutorial should guide user either to accept all default or custom project name when setting up environment. Here is what I entered:
```
project name: (pulumi-challenge) pulumi-challenge
project description: (A minimal Google Cloud TypeScript Pulumi program)
Created project 'pulumi-challenge'

Please enter your desired stack name.
To create a stack in an organization, use the format <org-name>/<stack-name> (e.g. `acmecorp/dev`).
stack name: (dev)
Created stack 'dev'

gcp:project: The Google Cloud project to deploy into:      
Saved config
```

File: themes/default/content/challenge/startup-in-a-box.md [hunk 5]
```diff
@@ -508,9 +513,410 @@ expect(await page.title()).toBe("Pulumi Challenge");
...
+  pulumi config set gcp:project <GCP_PROJECT_ID>
+
+  # or whatever your porjectid is
+  pulumi config set gcp:region europe-west1
+```
+
+#### Step 2. Creating Your First Resource
+
+Now that we have a base GCP project configured, we need to create our first resource. In this instance, we’ll create a new [GCS bucket](https://cloud.google.com/storage/docs/buckets) which will allow us to store our static website.
+
```
Comment (gangchen03) [inline themes/default/content/challenge/startup-in-a-box.md:563]: Need more instruction, following is a suggestion:
```
Create a new file named `create_gcs_bucket.ts` under the pulumi-challenge directory, and add the following code the typescript file.
```

File: themes/default/content/challenge/startup-in-a-box.md [hunk 6]
```diff
@@ -508,9 +513,410 @@ expect(await page.title()).toBe("Pulumi Challenge");
...
+});
+```
+
+#### Step 3. Working with Local Files
+
+Pulumi lets you use your favourite programming language to define your infrastructure. Today, we’re using TypeScript, which means we have access to the Node API. This includes discovering directories and files.
+
+From GCP we can use a synced folder to manage the files of the website.
+
+We need to add the synced-folder package from npm, to easiyl upload files to the Cloud Storage Bucket.
```
Comment (gangchen03) [inline themes/default/content/challenge/startup-in-a-box.md:590]: Typo from `easiyl` to `easily`.

File: themes/default/content/challenge/startup-in-a-box.md [hunk 7]
```diff
@@ -508,9 +513,410 @@ expect(await page.title()).toBe("Pulumi Challenge");
...
+
+```cli
+npm install @pulumi/synced-folder
+```
+
+Now with the synced-folder package installed we need to update our pulumi code.
```
Comment (gangchen03) [inline themes/default/content/challenge/startup-in-a-box.md]: should the user update the existing TS pulumi file created in step 2 or create a new file? If it is update existing file, will they append the following code or change certain lines? Instruction is not very clear, I did an add to the existing file I created earlier, right below the existing import statements. :

File: themes/default/content/challenge/startup-in-a-box.md [hunk 8]
```diff
@@ -508,9 +513,410 @@ expect(await page.title()).toBe("Pulumi Challenge");
...
+.banner p, .about p {
+    font-family: News Cycle;
+}
+```
+
+To make sure our styles display consistently across browsers, we also need to normalize some styles. [Copy normalize.css from GitHub](https://github.com/necolas/normalize.css/blob/master/normalize.css).
+
+#### Step 4. Creating a CDN
+
+Next, we want to front our Cloud Storage Bucket with a [Load Balancer](https://cloud.google.com/load-balancing/docs/https) and enable its [CDN capabilities](https://cloud.google.com/cdn/docs/overview).
```
Comment (gangchen03) [inline themes/default/content/challenge/startup-in-a-box.md]: The file name is confusing, seems like all the edits are happening in a single file `cdn-website.ts`? If so, doesn't make that much sense to ask user to continue editing the same file. We can break the code walk through into different sections if needed.

File: themes/default/content/challenge/startup-in-a-box.md [hunk 9]
```diff
@@ -508,9 +513,410 @@ expect(await page.title()).toBe("Pulumi Challenge");
...
+        });
+    }
+
+    get url(): pulumi.Output<string> {
+        return pulumi.interpolate`http://${this.ip.address}`;
+    }
+}
+```
+
+Now we can consume this! Awesome. Back in `pulumi-challenge/index.ts`, we now have this:
```
Comment (gangchen03) [inline themes/default/content/challenge/startup-in-a-box.md]: is this the first time I update the index.ts file. So, it is to `replace the index.ts with the following code`?

File: themes/default/content/challenge/startup-in-a-box.md [hunk 10]
```diff
@@ -508,9 +513,410 @@ expect(await page.title()).toBe("Pulumi Challenge");
...
+#### Step 6. Adding Another Provider
+
+Now that we have our website being delivered as fast as possible via our `CdnWebsite` component and Cloud CDN, how do we know that what we’ve deployed actually works? We could leverage a fantastic service, such as Checkly, to ensure our website passes some sanity checks.
+
+First, we need to add a new provider:
+
+```shell
+npm install @checkly/pulumi
+
+# API KEY: https://app.checklyhq.com/settings/account/api-keys
```
Comment (gangchen03) [inline themes/default/content/challenge/startup-in-a-box.md:880]: I would call out specifically how to create checkly API key by adding couple of sentence:
```
Please go to https://app.checklyhq.com/settings/account/api-keys to create a new API key. Then copy the key to clipboard.
```

File: themes/default/content/challenge/startup-in-a-box.md [hunk 11]
```diff
@@ -508,9 +513,410 @@ expect(await page.title()).toBe("Pulumi Challenge");
...
+pulumi config set checkly:apiKey --secret
+
+# AccountID: https://app.checklyhq.com/settings/account/general
+pulumi config set checkly:accountId
```
Comment (gangchen03) [inline themes/default/content/challenge/startup-in-a-box.md:884]: Better to have instruction on how to get the checkly accountID for someone new to checkly. Then ask them to paste to the command above.

File: themes/default/content/challenge/startup-in-a-box.md [hunk 12]
```diff
@@ -508,9 +513,410 @@ expect(await page.title()).toBe("Pulumi Challenge");
...
+```
+
+Next, we can use this in our code.
```
Comment (gangchen03) [inline themes/default/content/challenge/startup-in-a-box.md:887]: Which code? Assume `checkly-embed.js`?

Comment (gangchen03) [reply themes/default/content/challenge/startup-in-a-box.md:887]: Need to call out adding new file `checkly-embed.js` if so.

**Review by gangchen03 — COMMENTED** [review]
initial review comments and suggestion.

**Comment by gangchen03** [comment]
@timbohiatt failed to run the `pulumi up` with following error:
```
 pulumi:pulumi:Stack (pulumi-challenge-dev):
    error: Running program '/home/org_admin/pulumi/pulumi-challenge' failed with an unhandled exception:
    TSError: ⨯ Unable to compile TypeScript:
    index.ts(10,9): error TS2304: Cannot find name 'SIZE'.
```
Manually updated the index.ts file with an specific value for size as following:
```
const swag = new Swag("your-startup", {
  name: "YOUR NAME",
  email: "YOUR EMAIL",
  address: "YOUR ADDRESS",
  size: "M",

``` 
After the change, was able to run the pulumi up command successfully.

File: themes/default/content/challenge/startup-in-a-box.md (hunk 4 above)
Comment (timbohiatt) [reply themes/default/content/challenge/startup-in-a-box.md:551]: In line with the Pulumi owned and created content I will leave this to the user decide. Initially I thought about writing this or similar but after speaking with the Dev team or the Templates they leave some content knowledge "assumed" and they say following the on screen prompts is part of that assumption.

Comment (timbohiatt) [reply themes/default/content/challenge/startup-in-a-box.md:551]: I have mad a modification to include mention of following the prompts.

File: themes/default/content/challenge/startup-in-a-box.md (hunk 5 above)
Comment (timbohiatt) [reply themes/default/content/challenge/startup-in-a-box.md:563]: I have clarified by suggesting clearing the index.ts that comes withe gcp typescript index.ts file

File: themes/default/content/challenge/startup-in-a-box.md (hunk 9 above)
Comment (timbohiatt) [reply themes/default/content/challenge/startup-in-a-box.md]: I have worked through the text to ensure this should be clear going forward.

**Comment by timbohiatt** [comment]
@gangchen03 I have taken into consideration all of your suggestions and made significant modifications to the writing and also to the code. I have also retested the run. It should be good now. Thank you so much for the review.

**Pull request closed without merging**